paddle/framework/data_transform.cc

```diff
@@ -15,18 +15,58 @@ limitations under the License. */
 #include "paddle/framework/data_transform.h"
 
 #include "paddle/framework/data_device_transform.h"
+#include "paddle/framework/data_layout_transform.h"
 
 namespace paddle {
 namespace framework {
 
+static void free_tmp_tensor(const Tensor* in_ptr, const Tensor* tmp_ptr) {
+  if (in_ptr != tmp_ptr) {
+    delete tmp_ptr;
+  }
+}
+
 void DataTransform(const OpKernelType& expected_kernel_type,
                    const OpKernelType& kernel_type_for_var,
                    const Tensor& input_tensor, Tensor* out) {
+  // in_ptr is used to store the input of data transform function.
+  // it need to be deleted by free_tmp_tensor because ip_ptr may be
+  // equal to &input_tensor and input_tensor should not be deleted here.
+  const Tensor* in_ptr = &input_tensor;
+
+  // out_ptr is used to store the result of data transform function.
+  // when one transform is done, the in_ptr should be deleted and the out_ptr
+  // should be assigned to in_ptr for next data transformation.
+  Tensor* out_ptr = new Tensor();
```

Comment:
new and delete inside DataTransform looks like awful.  Furthermore, the memory optimizer cannot track these actions. 
How about create 3 Tensor(for each channel transform) in scope?

Reply:
Ok, let me think a better way to do this~